File: modules/java/main/src/main/java/com/mooltiverse/oss/nyx/services/github/GitHubRelease.java
/*
 * Copyright 2020 Mooltiverse
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.mooltiverse.oss.nyx.services.github;

import java.util.HashSet;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

import com.mooltiverse.oss.nyx.entities.Attachment;
import com.mooltiverse.oss.nyx.services.Release;

/**
 * A GitHub release.
 */
public class GitHubRelease extends GitHubEntity implements Release {
    /**
     * The internal set of release assets
     */
    private Set<Attachment> assets = null;

    /**
     * Creates the user object modelled by the given attributes.
     * 
     * @param api the reference to the API used to communicate with the remote end. Can't be {@code null}
     * @param attributes the map of attributes for this object. Can't be {@code null}
     * 
     * @throws NullPointerException if the given attributes map is {@code null}
     * @throws IllegalArgumentException if the map of attributes is empty
     */
    GitHubRelease(API api, Map<String, Object> attributes) {
        super(api, attributes);
    }

    /**
     * Creates the user object modelled by the given attributes.
     * 
     * @param api the reference to the API used to communicate with the remote end. Can't be {@code null}
     * @param attributes the map of attributes for this object. Can't be {@code null}
     * @param assets the set of assets attached to this release. It may be {@code null}
     * 
     * @throws NullPointerException if the given attributes map is {@code null}
     * @throws IllegalArgumentException if the map of attributes is empty
     */
    GitHubRelease(API api, Map<String, Object> attributes, Set<Attachment> assets) {
        super(api, attributes);
        this.assets = assets;
    }

    /**
     * Adds the given asset to the internal set of assets. The internal set of assets is initialized
     * in case it was still {@code null}.
     * 
     * @param asset the asset to add
     * 
     * @return this same object reference
     */
    public GitHubRelease addAsset(Attachment asset) {
        if (!Objects.isNull(asset)) {
            if (Objects.isNull(assets)) {
                assets = new HashSet<Attachment>();
            }
            assets.add(asset);
        }
            
        return this;
    }

    /**
     * Adds the given assets to the internal set of assets. The internal set of assets is initialized
     * in case it was still {@code null}.
     * 
     * @param assets the assets to add
     * 
     * @return this same object reference
     */
    public GitHubRelease addAssets(Set<Attachment> assets) {
        if (!Objects.isNull(assets)) {
            if (Objects.isNull(this.assets)) {
                this.assets = new HashSet<Attachment>();
            }
            this.assets.addAll(assets);
        }
            
        return this;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public Set<Attachment> getAssets() {
        return assets;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String getTitle() {
        return getAttributes().get("name").toString();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String getTag() {
        return getAttributes().get("tag_name").toString();
    }
}

File: src/go/nyx/services/github/github_release.go
/*
 * Copyright 2020 Mooltiverse
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package github

import (
	gh "github.com/google/go-github/github" // https://pkg.go.dev/github.com/google/go-github/github

	ent "github.com/mooltiverse/nyx/src/go/nyx/entities"
)

/*
A user for a remote GitHub release.
*/
type GitHubRelease struct {
	// The assets attached to the relese, or nil.
	assets []ent.Attachment

	// The release ID
	id int64

	// The tag the release refers to.
	tag string

	// The release title.
	title string
}

/*
Creates the user object modelled by the attributed from the given reference.

Arguments are as follows:

  - release the object to read the attributes from
*/
func newGitHubRelease(release gh.RepositoryRelease) *GitHubRelease {
	res := &GitHubRelease{}
	res.id = *release.ID
	res.tag = *release.TagName
	res.title = *release.Name
	return res
}

/*
Adds the given asset to the internal set of assets. The internal set of assets is initialized
in case it was still nil.

Arguments are as follows:

- asset the asset to add
*/
func (r *GitHubRelease) addAsset(asset ent.Attachment) *GitHubRelease {
	r.assets = append(r.assets, asset)

	return r
}

/*
Adds the given assets to the internal set of assets. The internal set of assets is initialized
in case it was still nil.

Arguments are as follows:

- assets the assets to add
*/
func (r *GitHubRelease) addAssets(assets []ent.Attachment) *GitHubRelease {
	r.assets = append(r.assets, assets...)

	return r
}

/*
Returns the assets attached to the relese, otherwise returns nil.
*/
func (r *GitHubRelease) GetAssets() []ent.Attachment {
	return r.assets
}

/*
Returns the release ID.
*/
func (r *GitHubRelease) GetID() int64 {
	return r.id
}

/*
Returns the tag the release refers to.
*/
func (r *GitHubRelease) GetTag() string {
	return r.tag
}

/*
Returns the release title.
*/
func (r *GitHubRelease) GetTitle() string {
	return r.title
}
